### PR TITLE
Check for missing WPA CRD before starting WPA controller loop

### PR DESCRIPTION
**Pull request by xornivore**
### What does this PR do?

Wait for WPA CRD infinitely if it is not registered before entering WPA controller loop. 
- Prevents frequent error logs from the Informer trying to list WPA resources.

This uses exponential backoff to increase delay between WPA CRD checks.

### Motivation

If you rollout the DCA without the CRD declared, you will see the following rather frequent log message:

```
E1001 16:14:05.575077       1 reflector.go:125] github.com/DataDog/watermarkpodautoscaler/pkg/client/informers/externalversions/factory.go:106: Failed to list *v1alpha1.WatermarkPodAutoscaler: the server could not find the requested resource (get watermarkpodautoscalers.datadoghq.com)
```

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/util/kubernetes/apiserver/wpa_controller.go [hunk 1]
```diff
@@ -38,8 +54,62 @@ func (h *AutoscalersController) RunWPA(stopCh <-chan struct{}) {
...
+	exp := &backoff.ExponentialBackOff{
+		InitialInterval:     crdCheckInitialInterval,
+		RandomizationFactor: backoff.DefaultRandomizationFactor,
+		Multiplier:          crdCheckMultiplier,
+		MaxInterval:         crdCheckMaxInterval,
+		MaxElapsedTime:      0,
+		Clock:               backoff.SystemClock,
+	}
+	exp.Reset()
+	_ = backoff.Retry(checkWPACRD(wpaClient), exp)
```
Comment (vboulineau) [inline pkg/util/kubernetes/apiserver/wpa_controller.go]: Is there a way to get the next "wait" time, so we can display it in the logs? Would be useful if people want to know if stuff is dead or just waiting for retry

Comment (xornivore) [reply pkg/util/kubernetes/apiserver/wpa_controller.go]: Ah yes, we could do that, though worth mentioning that we set the max of 	`crdCheckMaxInterval     = 5 * time.Minute` for this retrier.

**Review by vboulineau — DISMISSED** [review]

**Review by CharlyF — DISMISSED** [review]

File: pkg/util/kubernetes/apiserver/wpa_controller.go [hunk 2]
```diff
@@ -38,8 +55,69 @@ func (h *AutoscalersController) RunWPA(stopCh <-chan struct{}) {
...
+	reason := status.Status().Reason
+	details := status.Status().Details
+	return reason == v1.StatusReasonNotFound &&
+		details.Group == apis_v1alpha1.SchemeGroupVersion.Group &&
+		details.Kind == "watermarkpodautoscalers"
+}
+
+func checkWPACRD(wpaClient wpa_client.Interface) backoff.Operation {
+	check := func() error {
+		_, err := wpaClient.DatadoghqV1alpha1().WatermarkPodAutoscalers(v1.NamespaceAll).List(v1.ListOptions{})
```
Comment (clamoriniere) [inline pkg/util/kubernetes/apiserver/wpa_controller.go]: maybe you can use the `CustomResourceDefinition` client here.
something like:
```
client.ApiextensionsV1beta1().CustomResourceDefinitions().Get(...)
```

File: pkg/util/kubernetes/apiserver/wpa_controller.go [hunk 3]
```diff
@@ -38,8 +55,69 @@ func (h *AutoscalersController) RunWPA(stopCh <-chan struct{}) {
...
+		return err
+	}
+	return func() error {
+		return tryCheckWPACRD(check)
+	}
+}
+
+func waitForWPACRD(wpaClient wpa_client.Interface) {
+	exp := &backoff.ExponentialBackOff{
```
Comment (clamoriniere) [inline pkg/util/kubernetes/apiserver/wpa_controller.go]: did you look at kubernetes internal for backoff implementation: `wait.UntilWithContext `

```https://github.com/kubernetes/apimachinery/blob/9a93b3ad87694c17927348f74a0d8742ffa6221c/pkg/util/wait/wait.go#L96
```

**Review by clamoriniere — COMMENTED** [review]
looks great ! 
Maybe we can remove the new dependency by using internal kubernetes helper

File: pkg/util/kubernetes/apiserver/wpa_controller.go (hunk 2 above)
Comment (xornivore) [reply pkg/util/kubernetes/apiserver/wpa_controller.go]: Unfortunately the `clientSet` in WPA controller does not have `ApiextensionsV1beta()` API set.

File: pkg/util/kubernetes/apiserver/wpa_controller.go (hunk 3 above)
Comment (xornivore) [reply pkg/util/kubernetes/apiserver/wpa_controller.go]: Thank you @clamoriniere, I looked at the `ExponentialBackoff`, it's pretty nice but unfortunately it doesn't have a couple of convenient features:
a) it doesn't have a _clean way_ to enable infinite retries.
b) it doesn't have a way to retrieve the delay for the next attempt.

**Review by clamoriniere — APPROVED** [review]
👍

**Pull request merged**